modules/light-clients/07-tendermint/types/update_test.go

```diff
@@ -100,6 +105,70 @@ func (suite *TendermintTestSuite) TestCheckHeaderAndUpdateState() {
+				consensusState = types.NewConsensusState(suite.clientTime, commitmenttypes.NewMerkleRoot(suite.header.Header.GetAppHash()), suite.valsHash)
+				newHeader = suite.chainA.CreateTMClientHeader(chainID, int64(heightPlus1.RevisionHeight), height, suite.headerTime, suite.valSet, suite.valSet, signers)
+				currentTime = suite.now
+				ctx := suite.chainA.GetContext().WithBlockTime(currentTime)
+				// Store the header's consensus state in client store before UpdateClient call
+				suite.chainA.App.GetIBCKeeper().ClientKeeper.SetClientConsensusState(ctx, clientID, heightPlus1, newHeader.ConsensusState())
+			},
+			expFrozen: false,
+			expPass:   true,
+		},
+		{
+			name: "misbehaviour detection: header conflicts with existing consensus state",
+			setup: func(suite *TendermintTestSuite) {
+				clientState = types.NewClientState(chainID, types.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, heightPlus1, commitmenttypes.GetSDKSpecs(), upgradePath, false, false)
+				consensusState = types.NewConsensusState(suite.clientTime, commitmenttypes.NewMerkleRoot(suite.header.Header.GetAppHash()), suite.valsHash)
+				newHeader = suite.chainA.CreateTMClientHeader(chainID, int64(heightPlus1.RevisionHeight), height, suite.headerTime, suite.valSet, suite.valSet, signers)
+				currentTime = suite.now
+				ctx := suite.chainA.GetContext().WithBlockTime(currentTime)
+				// Change the consensus state of header and store in client store to create a conflict
+				conflictConsState := newHeader.ConsensusState()
+				conflictConsState.Root = commitmenttypes.NewMerkleRoot([]byte("conflicting apphash"))
+				suite.chainA.App.GetIBCKeeper().ClientKeeper.SetClientConsensusState(ctx, clientID, heightPlus1, conflictConsState)
+			},
+			expFrozen: true,
+			expPass:   true,
+		},
+		{
+			name: "misbehaviour detection: previous consensus state time is not before header time. time monotonicity violation",
+			setup: func(suite *TendermintTestSuite) {
+				clientState = types.NewClientState(chainID, types.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, height, commitmenttypes.GetSDKSpecs(), upgradePath, false, false)
+				// swap header and client time to create violation
+				consensusState = types.NewConsensusState(suite.clientTime, commitmenttypes.NewMerkleRoot(suite.header.Header.GetAppHash()), suite.valsHash)
+				newHeader = suite.chainA.CreateTMClientHeader(chainID, int64(heightPlus5.RevisionHeight), height, suite.headerTime, suite.valSet, suite.valSet, signers)
+				currentTime = suite.now
+				prevConsensusState := types.NewConsensusState(suite.headerTime, commitmenttypes.NewMerkleRoot(suite.header.Header.GetAppHash()), suite.valsHash)
+				ctx := suite.chainA.GetContext().WithBlockTime(currentTime)
+				suite.chainA.App.GetIBCKeeper().ClientKeeper.SetClientConsensusState(ctx, clientID, heightPlus1, prevConsensusState)
```

Comment:
it's hard to verify this test. There appears to be 3 times:
- client time
- header time
- suite.now

I'm not sure what is before or after without looking into the helper file. Can you add comments as to why this test functions correctly?

I'd assume header time to be after client time. The header is made with the header time, the consensus state is made with the client time, and the previous consensus state is made with header time and the context is set to suite.now. What is the consensus state used for?

Reply:
The consensus state with header time is just used to create the time violation. I create a monotonic time violation by setting a consensus state at a different height to be the same time as header.

I still need a valid consensus state in order to even pass the update. Note: the trusted height of header points to the valid consensus state at client time.

Documented to make things clearer.

Agreed that client time and header time can be confusing, but it's not just in this test case but throughout tendermint package. future testing pr should resolve this